fairseq2n/src/fairseq2n/data/sample_data_source.cc

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "fairseq2n/data/sample_data_source.h"
+
+#include <ATen/CPUGeneratorImpl.h>
+#include <ATen/Context.h>
+#include <ATen/Functions.h>
+
+#include <stdexcept>
+
+#include "fairseq2n/utils/tensor.h"
+
+namespace fairseq2n::detail {
+
+sample_data_source::sample_data_source(std::vector<data_pipeline> &&pipelines, std::vector<float> &&weights)
+    : pipelines_(std::move(pipelines))
+{
+    auto pipelines_count = pipelines_.size();
+
+    weights_ = make_tensor_from_vector(weights, { static_cast<std::int64_t>(pipelines_count) });
+    generator_ = at::globalContext().defaultGenerator(c10::DeviceType::CPU);
+}
+
+std::optional<data>
+sample_data_source::next()
+{
+    if (eod_)
+        return std::nullopt;
+
+    std::optional<data> output = pipelines_[next_index()].next();
+    if (!output)
+        eod_ = true;
+
+    return output;
+}
+
+void
+sample_data_source::reset()
+{
+    auto seed = generator_.current_seed();
+    generator_ = at::globalContext().defaultGenerator(c10::DeviceType::CPU);
+    generator_.set_current_seed(seed);
```

Comment:
Please don't do this. As a library we should never ever modify the default RNG. This can mess up the entire training flow. If the user wants repeatable reads from the pipeline, they should call `torch.manual_seed()`.

Reply:
can you explain this in bit more details, it's not really clear to me.
I thought that this would create a new generator just for this pipeline operator. Are you saying that `defaultGenerator` is just returning a pointer to a shared RNG ? Should we use a shared RNG, one for the op, or one for the full pipeline ?

Reply:
Yes, this just returns a handle to the shared (i.e. default) CPU RNG of the entire process. We definitely should not modify it as that will likely mess up other parts of users' code. In fact, I am not even entirely convinced that we should have a "seed" or RNG in the data pipeline logic at all. The typical use case is for users is to save/restore default RNGs (for CPU and CUDA) as part of checkpoint logic and that should be sufficient to restore randomness in data pipeline ops as well. I might be overseeing something though, let me know if you can think of a specific case where an explicit seed/RNG would be needed.

Reply:
In that case I agree that we should not have a seed in our pipeline. We can let the `at::Generator` handle it when the user calls `torch.manual_seed`